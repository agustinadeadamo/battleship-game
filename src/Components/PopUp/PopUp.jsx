/**
 * @desc Dependencies
 */
import React, { useEffect } from 'react';
import PropTypes from 'prop-types';

/**
 * @desc Styles
 */
import PopUpComponent from './style';

const PopUp = (props) => {
  // Destructuring props
  const {
    copy,
    show,
    hidePopUp,
  } = props;

  /**
   * Hook that is executed every time show is updates
   */
  useEffect(() => {
    let timer = null;

    // Validates if show is true
    if (show) {
      // Generates timeout that will hide pop up after 3 seconds
      timer = setTimeout(() => {
        hidePopUp();
      }, 1000);
    }

    // Clears timeout
    return () => clearTimeout(timer);
  }, [show]);

  return (
    <PopUpComponent data-testid="popUp-component" className={`${show ? 'show' : 'hide'}`} show={show}>
      {copy}
    </PopUpComponent>
  );
};

PopUp.propTypes = {
  copy: PropTypes.string.isRequired,
  show: PropTypes.bool.isRequired,
  hidePopUp: PropTypes.func.isRequired,
};

export default PopUp;
