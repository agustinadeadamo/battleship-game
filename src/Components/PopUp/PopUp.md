# Input Component
Pop up that is shown with a notification

## Hooks

+ **useEffect** it's executed every time the prop "show" changes, when it's value is true, it executes the settimeout that indicates that pop should be hidden in one second, it's indicated throught the function "hidePop" that is send to it's parent

## Props
+ **copy**  String
+ **show**  Boolean
+ **hidePopUp**  Function