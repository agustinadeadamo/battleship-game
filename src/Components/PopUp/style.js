/**
 * @desc Dependencies
 */
import styled from 'styled-components';
import Colors from '../../MainStyles/Variables';

/**
 * @desc Variabled
 */

const PopUpComponent = styled.div`
  width: 100%;
  max-width: 300px;
  height: 50px;
  position: fixed;
  left: 0;
  right: 0;
  margin-left: auto;
  margin-right: auto;
  color: ${Colors.white};
  font-size: 13px;
  background: ${Colors.green};
  font-weight: 600;
  border-radius: 6px;
  text-align: center;
  display: flex;
  justify-content: center;
  align-items: center;

  &.show {
    top: 0; 
    transition: all 0.5s ease 0s; 
    -webkit-transition: all 0,5s ease 0s; 
  }

  &.hide {
    top: -50%; transition: all 0.5s ease 0s; -webkit-transition: all 0.5s ease 0s;
  }
`;

export default PopUpComponent;
