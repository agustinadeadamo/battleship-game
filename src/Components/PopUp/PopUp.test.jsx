/**
 * @desc Dependencies
 */
import { create, act } from 'react-test-renderer';
import React from 'react';
import { render, screen } from '@testing-library/react';
import {
  describe, expect, it, jest,
} from '@jest/globals';

/**
 * @desc Components
 */
import PopUp from './PopUp';

describe('[PopUp Component]', () => {
  it('Renders correctly', () => {
    const { queryByTestId } = render(<PopUp show copy="test" hidePopUp={() => {}} />);
    expect(queryByTestId('popUp-component')).toBeTruthy();
  });

  it('Matches snapshot', () => {
    const input = create(<PopUp show copy="test" hidePopUp={() => {}} />).toJSON;
    expect(input).toMatchSnapshot();
  });

  it('Hides pop up after three seconds', () => {
    // Mocks timers
    jest.useFakeTimers();
    // Renders with show value equal to true
    const { rerender } = render(<PopUp show copy="test" hidePopUp={() => {}} />);
    // Changes prop
    rerender(<PopUp copy="test" show={false} hidePopUp={() => {}} />);
    // Executes setTtimeoute
    act(() => {
      jest.runAllTimers(); // trigger setTimeout
    });
    // Validates that class has changes
    expect(screen.queryByTestId('popUp-component')).toHaveClass('hide');
  });
});
