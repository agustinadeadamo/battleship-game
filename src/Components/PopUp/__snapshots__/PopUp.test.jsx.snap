// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`[PopUp Component] Matches snapshot 1`] = `[Function]`;
